README.md
TwitterX v1.3.1
===============

This uses twitteroauth: https://github.com/abraham/twitteroauth

This package loads Twitter feeds using the new (and very annoying) Twitter 1.1 API. You will need to create a Twitter app here: 

https://dev.twitter.com/apps/new

Once you've created your new application you will need to generate tokens and use those tokens when you call the TwitterX snippet which will load your statuses:



Snippet TwitterX
----------------

This should be called uncached if you are using :ago in the chunk otherwise it can be cached.

[[!TwitterX? 
    &twitter_consumer_key=`aaaa`
    &twitter_consumer_secret=`bbbb` 
    &twitter_access_token=`cccc` 
    &twitter_access_token_secret=`dddd`
    &limit=`4`
    &timeline=`user_timeline`
    &chunk=`TwitterXTpl`
    &cache=`7200`
    &screen_name=`Qodo,Microsoft`
    &include_rts=`1`
]]

twitter_consumer_key - your twitter consumer token (REQUIRED)
twitter_consumer_secret - your twitter consumer secret (REQUIRED)
twitter_access_token - your twitter access token (REQUIRED)
twitter_access_token_secret - your twitter access token secret (REQUIRED)
limit - limit how many statuses to display (default: 5)
timeline - which twitter timeline to load (default: user_timeline, retweeted_by_me etc)
chunk - which chunk to load when rendering the statuses (default: TwitterXTpl)
cache - how many seconds to cache the twitter data feed (default: 7200)
screen_name - which user you want to load
include_rts - should this include retweets (default: 1)
cache_id - unique ID for caching in case you want to view multiple feeds or different feeds (default: TwitterX_PAGEID)
toPlaceholder - a placeholder ID if you want to use content as a placeholder instead of outputting directly (default: '')
toPlaceholderPrefix - if you want to prefix the placeholder values. E.g. 'twitterx' would create placeholders like [[*twitterx.text]] (default: '')
slug - for when loading a twitter list, you must specify a screen_name and a slug (default: '')



Loading timelines
-----------------

As the 1.1 API is more restrictive the tweets available have changed. The snippet defaults to user_time line but you can load any of these:

public_timeline
public_timeline
friends_timeline
user_timeline
mentions
retweets_of_me
favourites



Loading lists ** New as of 1.3 **
---------------------------------

You can now load lists by using the following:

    &timeline=`lists/statuses`
    &slug=`NAME_OF_YOUR_SLUG`
    &screen_name=`SCREEN_NAME_OF_LIST_OWNER`

An example of this would be the MODX List: https://twitter.com/modx/the-modx-team

    &timeline=`lists/statuses`
    &slug=`the-modx-team`
    &screen_name=`modx`



Searching Twitter
-----------------

TwitterX now supports basic Twitter searches using the &search parameter:

&search=`MODX`



Snippet TwitterXFormat
----------------------

Use this snippet in your chunk placeholders to format the status text and automatically link any search, usernames or topics:

[[+text:TwitterXFormat]]



Chunk options
-------------

The package comes with a chunk for displaying the statuses called 'TwitterXTpl'. You can customise this by using the following placeholders:

created_at - date status was created
source - source of the Tweet (application like web, iOS etc)
id - status id on timeline
id_str - status id on Twitter (twitter.com/user/statuses/id_str)
text - status main text
name - Twitter name
screen_name - Twitter username
profile_image_url - Twitter avatar image url for this user
location - This users location
url - This users URL (if specified)
description - This users profile information



Retweets (where applicable)
--------------------------

retweet_created_at - date status was created
retweet_source - source of the Tweet (application like web, iOS etc)
retweet_id - status id on timeline
retweet_id_str - status id on Twitter (twitter.com/user/statuses/id_str)
retweet_text - status main text
retweet_name - Twitter name
retweet_screen_name - Twitter username
retweet_profile_image_url - Twitter avatar image url for this user
retweet_location - This users location
retweet_url - This users URL (if specified)
retweet_description - This users profile information



Further info
------------

For information and support, check out my blog:

http://www.qodo.co.uk/blog/twitterx-a-new-modx-extra-for-pulling-in-twitter-feeds-using-api-1.1/

Created by Stewart Orr @ Qodo Ltd (http://www.qodo.co.uk).
Contributers: @sepiariver (http://www.sepiariver.ca), @OostDesign (http://www.oostdesign.com/), @scottborys (http://scottborys.com/)

docs/changelog.txt
Changelog for TwitterX package. 

TwitterX 0.8 - 21/01/2013
===================================== 
- Added option include_rts to choose whether to include retweets.
- Added options cache_id and toPlaceholder thanks to @sepiariver to allow for placeholder output and unique caching.
- Fix for viewing Twitter favourites instead of regular statuses (thanks @sepiariver).
- Added a few more comments.

TwitterX 0.7 - 01/12/2012
===================================== 
- Fixed spelling mistake in error message
- Added test for CURL as it's needed by twitter auth
- Corrected parameter 'twitter_consumer_token_secret' which should have been 'twitter_access_token_secret' with support for the old

TwitterX 0.6 - 01/12/2012
===================================== 
- Added option to view other users statuses using &screen_name=`Qodo`

TwitterX 0.5 - 29/11/2012
===================================== 
- Changed the way we limit the number of tweets - using twitter 'count' parameter instead of a limited loop.

TwitterX 0.4 - 15/11/2012
===================================== 
- Bug fix for reporting errors (thanks Jeremy)

TwitterX 0.3
===================================== 
- Added in caching of twitter data to protect against over using twitter limit
- Added id_str available to chunk (thanks Laurent Mignot)
- Updated to check for the twitteroauth class first
- Corrected error in documentation (feed parameter should have been timeline)

TwitterX 0.2
===================================== 
- First version

TwitterX.snippet.php
<?php
/**
 * TwitterX
 *
 * This package loads Twitter feeds using the new (and very annoying) Twitter
 * 1.1 API. You will need to create a Twitter app and get the keys and tokens
 * by creating a new app here: https://dev.twitter.com/apps/new
 *
 * This uses twitteroauth: https://github.com/abraham/twitteroauth
 *
 * TwitterX is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option) any
 * later version.
 *
 * TwitterX is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 
 * @author Stewart Orr @ Qodo Ltd <stewart@qodo.co.uk>
 * @version 0.9
 * @copyright Copyright 2012 by Qodo Ltd
 * With thanks to Sepiariver http://www.sepiariver.ca/
 */

// Twitter API keys and secrets
$twitter_consumer_key          = isset($twitter_consumer_key) ? $twitter_consumer_key : FALSE ;
$twitter_consumer_secret       = isset($twitter_consumer_secret) ? $twitter_consumer_secret : FALSE ;
$twitter_access_token          = isset($twitter_access_token) ? $twitter_access_token : FALSE ;
$twitter_access_token_secret   = isset($twitter_access_token_secret) ? $twitter_access_token_secret : FALSE ;

// Other options
$limit = isset($limit) ? $limit : 5 ;
$chunk = isset($chunk) ? $chunk : 'TwitterXTpl' ;
$timeline = isset($timeline) ? $timeline : 'user_timeline' ;
$cache = isset($cache) ? $cache : 7200 ;
$screen_name = isset($screen_name ) ? $screen_name : '' ;
$include_rts = isset($include_rts) ? $include_rts : 1 ;
$cache_id = isset($cache_id) ? $cache_id : 'TwitterX' ;
$toPlaceholder = isset($toPlaceholder) ? $toPlaceholder : '' ;

// Here we support an old error where the parameter was incorrect.
if (isset($twitter_consumer_token_secret)) {
	$twitter_access_token_secret = isset($twitter_consumer_token_secret) ? $twitter_consumer_token_secret : FALSE ;
}

// Function to compare tweets based on their creation time
if (!function_exists('compareTweetsByDate')) {
	function compareTweetsByDate($a, $b) {
		$time_a = strtotime($a->created_at);
		$time_b = strtotime($b->created_at);
		
		if ($time_a == $time_b) {
		  return 0;
		}
		return ($time_a > $time_b) ? -1 : 1;
	}
}

// HTML output 
$output = '';

//**************************************************************************

// If they haven't specified the required Twitter keys, we cannot continue...
if (!$twitter_consumer_key || !$twitter_consumer_secret || !$twitter_access_token || !$twitter_access_token_secret) {

	echo "<strong>TwitterX Error:</strong> Could not load TwitterX as required values were not passed.";

} else {

	// Test for required function(s)
	if (!function_exists('curl_init')) {
			echo "<strong>TwitterX Error:</strong> cURL functions do not exist, cannot continue.";	
	} else {

		// Try loading the data from MODX cache first
		$json = $modx->cacheManager->get($cache_id . '_' .  $modx->resource->id); // Added ability to set custom cache IDs
		
		if (!$json) { 

			// Load the TwitterOAuth lib required if not exists
			if (!class_exists('TwitterOAuth')) {
				require_once $modx->getOption('core_path').'components/twitterx/twitteroauth/twitteroauth/twitteroauth.php';
			}
			// Create new twitteroauth
			$twitteroauth = new TwitterOAuth($twitter_consumer_key, $twitter_consumer_secret, $twitter_access_token, $twitter_access_token_secret);

			// We want to use JSON format
			$twitteroauth->format = 'json';
			$twitteroauth->decode_json = FALSE;

			// Request statuses with optinal parameters
			$options = array(
				'count' => $limit+1,
				'include_rts' => $include_rts
			);

			// If we are viewing favourites or regular statuses
			if ($timeline != 'favorites') {
				$timeline = 'statuses/' . $timeline;
			}

			$tweets = array();
			// If we have one or multiple screen names
			if ($screen_name != '') {

				// Collect screen_names
				$screen_name_array = preg_split("/,/", $screen_name, -1, PREG_SPLIT_NO_EMPTY);

				if (count($screen_name_array) >= 1) {
					// Get timeline for every screen name
					foreach ($screen_name_array as $sn) {
						$options['screen_name'] = $sn;
						$json_part = $twitteroauth->get($timeline, $options);

						// No error while loading timeline
						if (!isset($json_part->error)) {
							$tweets = array_merge($tweets, json_decode($json_part));
						}
					}
				}
			}

			// Sort mixed tweets of different users
			usort($tweets, 'compareTweetsByDate');

			// Limit the combined result
			$tweets = array_slice($tweets, 0, $limit);

			// Some entries loaded? Save to MODX Cache
			if (count($tweets) > 0) {
				$json = json_encode($tweets);
				$modx->cacheManager->set($cache_id . '_' .  $modx->resource->id, $json, $cache);
			}
	
		}
	
		// Decode this now that we have used it above in the cache
		$json = json_decode($json);
	
		// If there any errors from Twitter, output them...
		if (isset($json->error)) {
			
			echo "<strong>TwitterX Error:</strong> Could not load TwitterX Twitter responded with the error '" . $json->error . "'.";
			
		} else {

			// Any tweets present?
			if (is_array($json) && count($json) > 0) {

				// For each result, output it
				foreach ($json as $j) {
					
					// Get placerholder values
					$placeholders = array(
						'created_at' => $j->created_at,
						'source' => $j->source,
						'id' => $j->id,
						'id_str' => $j->id_str,
						'text' => $j->text,
						'name' => $j->user->name,
						'screen_name' => $j->user->screen_name,
						'profile_image_url' => $j->user->profile_image_url,
						'location' => $j->user->location,
						'url' => $j->user->url,
						'description' => $j->user->description,
					);
					// If this is a retweet, create placeholders for this too
					if (isset($j->retweeted_status)) {
						$placeholders = array_merge($placeholders, array(
							'retweet_count' => $j->retweeted_status->retweet_count,
							'retweet_created_at' => $j->retweeted_status->created_at,
							'retweet_source' => $j->retweeted_status->source,
							'retweet_id' => $j->retweeted_status->id,
							'retweet_id_str' => $j->retweeted_status->id_str,
							'retweet_text' => $j->retweeted_status->text,
							'retweet_name' => $j->retweeted_status->user->name,
							'retweet_screen_name' => $j->retweeted_status->user->screen_name,
							'retweet_profile_image_url' => $j->retweeted_status->user->profile_image_url,
							'retweet_location' => $j->retweeted_status->user->location,
							'retweet_url' => $j->retweeted_status->user->url,
							'retweet_description' => $j->retweeted_status->user->description,
							)
						);
					}
					// Parse chunk passing values
					$output .= $modx->getChunk($chunk, $placeholders); // Concatenate to output variable
				}
			}

			// Added option to output to placeholder
			if ($toPlaceholder != '') {
				$modx->setPlaceholder($toPlaceholder, $output);
			} else {
				return $output;
			}
		}
	}
}